README.md

```diff
@@ -29,14 +33,75 @@ This package is developed by [valantic CEC Schweiz](https://www.valantic.com/en/
+methods are already implemented in the `AbstractDocument`, so it is recommended to use that one as a base class. Then
+only two methods need to be implemented:
+
+- `getType` is either asset, document, data object or variant. The `DocumentInterface` provides constants for them,
+  which can be returned in your implementation.
+- `getSubType` is very useful for data objects, since it allows to define what kind of data object this document is
+  about. It is best to use the `::class` constant of the data object.
+
+Mind that this document is not what is actually indexed, which is a `IndexDocument` (explained later). The document just
+described is completely separated from an index, which allows it to be reused in many different indices.
+
+See the [`ProductDocument` provided in the
+example](docs/example/src/AppBundle/Elasticsearch/Document/ProductDocument.php) for more details.
+
+### Define a mapping from element to document
+
+**The `IndexDocumentInterface` is mainly responsible for the mapping from a Pimcore element to document stored in an
+Elasticsearch index.** This class usually inherits from the corresponding `DocumentInterface` implementation, since many
+methods from that implementation are reusable.
+
+There are two traits that help a lot with implementing that interface:
+
+- The `ListingTrait` provides an implementation returning a Pimcore `Listing` class for the given element defined in the
+  `DocumentInterface`. This listing will be used to load data to store in the index.
+- The `DataObjectNormalizerTrait` helps with creating a mapping from a Pimcore element to an index by providing a few
+  helper methods.
+
+While the `ListingTrait` is just used and does not need any further effort, the `DataObjectNormalizerTrait` does not do
+anything on its own. It is used in the `IndexDocumentInterface::getNormalized` method, which creates an index
+representation of the Pimcore element. The `getNormalized` method returns the associative array to be indexed. The
+`DataObjectNormalizerTrait` helps in this process with the following methods, all of which take the Pimcore element and
+an array describing the mapping:
+
+- `plainAttributes` is for "normal" attributes. Based on the mapping the index will just be filled with the value in
+  these properties.
+- `localizedAttributes` is for localized Pimcore attributes. They will be stored in a `localized` field in the index
+  with all languages as children.
+- `relationAttributes` allow to store just a reference, i.e. the ID of a Pimcore element, instead of the entire object
+  in the index.
+
+The mapping can either be just a value array, in which case the Pimcore element's property will be indexed using the
+same name or a key-value pair if the property should be named differently in the index. **If a key-value pair is used,
+it is also possible to pass a function retrieving the Pimcore element and returning an arbitraty array.** This is very
```

Comment:
I think we really should add an example of this in the `examples` directory. It is very powerful, so it would be a shame if people do not end up using it.